### PR TITLE
Extend ICollection<T> and IList<T> handling in LINQ All/Any/First methods

### PR DESCRIPTION
**Pull request by UselessToucan**

### CODE REVIEW

File: src/System.Linq/src/System/Linq/AnyAll.cs [hunk 1]
```diff
@@ -15,6 +15,11 @@ public static bool Any<TSource>(this IEnumerable<TSource> source)
                 throw Error.ArgumentNull(nameof(source));
             }
 
+            if (source is ICollection<TSource> collection)
```
[the remark this answers is not in the document]
Comment (UselessToucan) [reply src/System.Linq/src/System/Linq/AnyAll.cs]: No, I don't have performance data to show the impact on both the success and failure cases.

[the remark this answers is not in the document]
Comment (khellang) [reply src/System.Linq/src/System/Linq/AnyAll.cs]: https://github.com/dotnet/coreclr/issues/603 might change those numbers, though.

[the remark this answers is not in the document]
Comment (OmarTawfik) [reply src/System.Linq/src/System/Linq/AnyAll.cs]: +1 @stephentoub 
We need to see perf numbers improvements for this change, and it needs to compare that to how much other types regressed.

**Comment by OmarTawfik** [comment]
@UselessToucan do you think you'll have the time to look into the feedback/perf measurement?

**Comment by UselessToucan** [comment]
@OmarTawfik I'm sorry, but I do not have suitable performance testing experience. Most likely I won't be able to provide correct performance data. :(

**Comment by stephentoub** [comment]
@OmarTawfik, what do you want to do with this PR?

**Comment by OmarTawfik** [comment]
I'll close for now as we cannot merge such changes without performance measurements.
We can reactive in the future if more info became available.

**Pull request closed without merging**